microsoft/testsuites/nested/common.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from typing import List, Optional
+
+from lisa import RemoteNode
+from lisa.schema import Node
+from lisa.tools import Qemu, Wget
+from lisa.tools.df import Df, PartitionInfo
+from lisa.util import SkippedException
+
+NESTED_VM_IMAGE_NAME = "image.qcow2"
+NESTED_VM_TEST_FILE_NAME = "message.txt"
+NESTED_VM_TEST_FILE_CONTENT = "Message from L1 vm!!"
+NESTED_VM_TEST_PUBLIC_FILE_URL = "http://www.github.com"
+NESTED_VM_REQUIRED_DISK_SIZE_IN_GB = 6
+
+
+def connect_nested_vm(
+    host: RemoteNode,
+    guest_username: str,
+    guest_password: str,
+    guest_port: int,
+    guest_image_url: str,
+    image_name: str = NESTED_VM_IMAGE_NAME,
+    image_size: int = NESTED_VM_REQUIRED_DISK_SIZE_IN_GB,
+    disks: Optional[List[str]] = None,
+) -> RemoteNode:
+    image_folder_path = get_partition_for_nested_image(host, image_size)
+
+    host.tools[Wget].get(
+        url=guest_image_url,
+        file_path=image_folder_path,
+        filename=image_name,
+        sudo=True,
+    )
+
+    # start nested vm
+    host.tools[Qemu].create_vm(guest_port, f"{image_folder_path}/{image_name}", disks)
+
+    # setup connection to nested vm
+    nested_vm = RemoteNode(Node(name="L2-vm"), 0, "L2-vm")
+    nested_vm.set_connection_info(
+        public_address=host.public_address,
+        username=guest_username,
+        password=guest_password,
+        public_port=guest_port,
+        port=guest_port,
+    )
+
+    return nested_vm
+
+
+def get_partition_for_nested_image(node: RemoteNode, size: int) -> str:
+    home_partition = node.tools[Df].get_partition_by_mountpoint("/home")
```

Comment:
call once on `node.tools[Df]` and assign it to a variable for the second use.

Reply:
The force run is false in the function call. So, it is internally one single call to df

Reply:
You can add force_run to the arguments, and give the False as default value. If the code writes like this way, it hides the reason, and may be "improved" in future, and leave a bug here.

Reply:
So, it's good to use only one `df = node.tools[Df]`